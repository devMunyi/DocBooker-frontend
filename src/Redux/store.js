import { configureStore, applyMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import doctorReducer from './reducers/doctors';
import reservationReducer from './reducers/reservations';
import selectedDoctorReducer from './reducers/selectedDoctor';
import allDoctors from './reducers/allDoctors';
import user from './reducers/user';


export default configureStore(
  {
    reducer: {
      doctor: doctorReducer,
      reservations: reservationReducer,
      selectedDoctor: selectedDoctorReducer,
      doctors: allDoctors,
      user,
    },
  },
  applyMiddleware(logger),
);
