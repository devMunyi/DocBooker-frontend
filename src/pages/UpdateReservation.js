import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import axios from 'axios';
import { DatePicker } from 'antd';
import { toast } from 'react-toastify';
import moment from 'moment/moment';
import dayjs from 'dayjs';

function UpdateReservation() {
  const navigate = useNavigate();

  const userId = sessionStorage.getItem('user')
    ? JSON.parse(sessionStorage.getItem('user')).id
    : null;

  if (!userId === null) {
    navigate('/login');
  }

  const [isProcessing, setIsProcessing] = useState(false);
  const [doctorId, setDoctorId] = useState(
    JSON.parse(sessionStorage.getItem('reservation')) ? JSON.parse(sessionStorage.getItem('reservation')).doctorid : null,
  );
  const [date, setDate] = useState(
    JSON.parse(sessionStorage.getItem('reservation')) ? JSON.parse(sessionStorage.getItem('reservation')).date : null,
  );

  const reservation = JSON.parse(sessionStorage.getItem('reservation'));
  const doctors = JSON.parse(sessionStorage.getItem('doctors'));

  const handleDateChange = (value, dateString) => {
    setDate(dateString);
  };

  const handleDoctorChange = (event) => {
    const selectedDoctorId = event.target.value;
    setDoctorId(selectedDoctorId);
  };

  const handleReservationUpdate = async () => {
    if (date === '') {
      toast.error('Date cannot be empty');
      return;
    }

    if (date.length !== 10) {
      toast.error('Invalid date input');
      return;
    }

    try {
      setIsProcessing(true);
      const response = await axios.put(
        `${process.env.REACT_APP_API_BASE_URL}/users/${userId}/doctors/${doctorId}/reservations/${reservation.id}`,
        {
          reservation: {
            user_id: userId,
            doctor_id: doctorId,
            date,
            details: 'Some details',
          },
        },
        {
          headers: {
            'Content-Type': 'application/json',
          },
        },
      );

      const { data, status } = response;
      const { id } = data;
      if (status === 200) {
        toast.success('Reservation updated successfully!');
        setTimeout(() => {
          navigate(`/reservations/${doctorId}/${id}`);
        }, 2500);
      } else {
        throw new Error(data.message || 'Failed to update reservation');
      }
    } catch (error) {
      toast.error(error.message || 'Failed to update reservation');
    } finally {
      setIsProcessing(false);
    }
  };

  return (
    <div id="reserve-page" className="reservePage">
      <div className="container-fluid reservePage">
        <img
          src={`${reservation?.doctorpicture}`}
          alt="Doctor"
          style={{ width: '100%', height: '100vh' }}
        />
        <div className="centered">
          <h1 className="doc-name pb-2">{`Dr ${reservation?.doctorname} - ${reservation?.specialization}`}</h1>
          <p>
            Lorem, ipsum dolor sit amet consectetur adipisicing elit. Minus
            provident et expedita perferendis dolore perspiciatis odio.
            Dignissimos, in deleniti pariatur alias odit explicabo vitae libero
            provident est placeat consectetur hic?
          </p>
          <div className="d-flex gap-3 justify-content-center">
            <select
              style={{ width: '27%' }}
              id="doctor-select"
              value={doctorId}
              onChange={handleDoctorChange}
              className="rounded-0"
            >
              {doctors && doctors.map((doctor) => (
                <option key={doctor.id} value={doctor.id}>
                  {doctor.name}
                </option>
              ))}
            </select>
            <DatePicker
              defaultValue={dayjs(reservation && reservation.date, 'YYYY-MM-DD')}
              format="YYYY-MM-DD"
              onChange={handleDateChange}
              className="date-picker rounded-0"
              disabledDate={(current) => current && current.valueOf() < moment().subtract(1, 'days')}
            />
          </div>

          <div className="d-flex gap-3 justify-content-center mt-4">
            <Link
              to={`/reservations/${reservation && reservation.doctorid}/${reservation && reservation.id}`}
              className="btn btn-outline-light pb-0"
            >
              Go Back
            </Link>
            {isProcessing ? (
              <button className="btn btn-outline-light" type="button" disabled>
                <span
                  className="spinner-border spinner-border-sm"
                  role="status"
                  aria-hidden="true"
                />
                <span className="visually-hidden">Loading...</span>
              </button>
            ) : (
              <button
                type="submit"
                onClick={handleReservationUpdate}
                className="btn btn-outline-light"
              >
                Update
              </button>
            )}
          </div>
        </div>
      </div>
    </div>
  );
}

export default UpdateReservation;
