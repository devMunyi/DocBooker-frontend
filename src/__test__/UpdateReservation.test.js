import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import store from '../Redux/store';
import UpdateReservation from '../pages/UpdateReservation';

it('UpdateReservation component renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <UpdateReservation />
        </Router>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
