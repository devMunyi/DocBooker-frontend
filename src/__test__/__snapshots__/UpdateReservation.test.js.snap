// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateReservation component renders correctly 1`] = `
<div
  className="reservePage"
  id="reserve-page"
>
  <div
    className="container-fluid reservePage"
  >
    <img
      alt="Doctor"
      src="undefined"
      style={
        Object {
          "height": "100vh",
          "width": "100%",
        }
      }
    />
    <div
      className="centered"
    >
      <h1
        className="doc-name pb-2"
      >
        Dr undefined - undefined
      </h1>
      <p>
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Minus provident et expedita perferendis dolore perspiciatis odio. Dignissimos, in deleniti pariatur alias odit explicabo vitae libero provident est placeat consectetur hic?
      </p>
      <div
        className="d-flex gap-3 justify-content-center"
      >
        <select
          className="rounded-0"
          id="doctor-select"
          onChange={[Function]}
          style={
            Object {
              "width": "27%",
            }
          }
          value={null}
        />
        <div
          className="ant-picker css-dev-only-do-not-override-10ed4xt date-picker rounded-0"
          onClick={[Function]}
        >
          <div
            className="ant-picker-input"
          >
            <input
              autoComplete="off"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              placeholder="Select date"
              readOnly={true}
              size={12}
              title=""
              value=""
            />
            <span
              className="ant-picker-suffix"
            >
              <span
                aria-label="calendar"
                className="anticon anticon-calendar"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="calendar"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                  />
                </svg>
              </span>
            </span>
            <span
              className="ant-picker-clear"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              role="button"
            >
              <span
                aria-label="close-circle"
                className="anticon anticon-close-circle"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="close-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                  />
                </svg>
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        className="d-flex gap-3 justify-content-center mt-4"
      >
        <a
          className="btn btn-outline-light pb-0"
          href="/reservations/null/null"
          onClick={[Function]}
        >
          Go Back
        </a>
        <button
          className="btn btn-outline-light"
          onClick={[Function]}
          type="submit"
        >
          Update
        </button>
      </div>
    </div>
  </div>
</div>
`;
